In [1]:
import numpy as np
import time

In [2]:
X=[0.5,2.5]
Y=[0.2,0.9]


In [3]:
def sigmoid(w,b,x):
    return 1/(1+(np.exp(-(w*x-b))))

In [4]:
def gradW(w,b,x,y):
    fx = sigmoid(w,b,x)
    return (fx - y) * (fx) * (1 - fx) * x

In [5]:
def gradB(w,b,x,y):
    fx = sigmoid(w,b,x)
    return (fx - y) * (fx) * (1 - fx)

In [6]:
def lossFuction(w,b):
    error = 0
    for (dx,dy) in zip(X,Y):
        fx = sigmoid(w,b,dx)
        error += 0.5 * (fx - dy)**2
    return error

In [12]:
def moumentum_grad(w,b,learning_rate,epochs):
    seconds = time.time()
    gama = 0.1
    w_pre,b_pre = 0,0
    for i in range(epochs):
        dw,db = 0,0
        for (x,y) in zip(X,Y):
            dw += gradW(w,b,x,y)
            db += gradB(w,b,x,y)
        w = w - ((gama * w_pre) +(learning_rate * dw))
        b = b - ((gama * b_pre) +(learning_rate * db))
        w_pre = w
        b_pre = b
    print(lossFuction(w,b))
    print("Seconds since epoch =", time.time() - seconds)

In [13]:
def nesterov_grad(w,b,learning_rate,epochs):
    seconds = time.time()
    gama = 0.1
    w_pre,b_pre = 0,0
    for i in range(epochs):
        dw,db = 0,0
        w = w - (gama * w_pre)
        b = b - (gama * b_pre)
        for (x,y) in zip(X,Y):
            dw += gradW(w,b,x,y)
            db += gradB(w,b,x,y)
        w = w - ((gama * w_pre) +(learning_rate * dw))
        b = b - ((gama * b_pre) +(learning_rate * db))
        w_pre = ((gama * w_pre) +(learning_rate * dw))
        b_pre = ((gama * b_pre) +(learning_rate * db))
    print(lossFuction(w,b))
    print("Seconds since epoch =", time.time() - seconds)

In [39]:
moumentum_grad(2,2,0.1,20)
print("--------------------------");
nesterov_grad(2,2,0.1,20)

0.07307142628533342
Seconds since epoch = 0.0005400180816650391
--------------------------
0.10476153557145784
Seconds since epoch = 0.0002541542053222656


<table>
    <tr>
        <th>Weight</th>
        <th>Bais</th>
        <th>Learning rate</th>
        <th>epoch</th>
        <th>Momentum-Gradient-Time</th>
        <th>Nesterov-Gradient-Time</th>
        <th>Momentum-Gradient-Loss</th>
        <th>Nesterov-Gradient-Loss</th>
        <th></th>
    </tr>
    <tr>
        <td>-2</td>
        <td>-2</td>
        <td>0.1</td>
        <td>10</td>
        <td> 0.00018978118896484375</td>
        <td>0.00013637542724609375</td>
        <td>0.26750690145032485</td>
        <td>0.1552378903673416</td>
    </tr>
      <tr>
        <td>-2</td>
        <td>-2</td>
        <td>0.1</td>
        <td>20</td>
        <td>0.0005130767822265625</td>
        <td> 0.00048732757568359375</td>
        <td>0.13669806796427464</td>
        <td>0.11132566037174835</td>
    </tr> 
    <tr>
        <td>2</td>
        <td>2</td>
        <td>0.1</td>
        <td>10</td>
        <td>0.00029468536376953125</td>
        <td>0.0001437664031982422</td>
        <td>0.030880688040618033</td>
        <td>0.07616201599640063</td>
    </tr>
    <tr>
        <td>2</td>
        <td>2</td>
        <td>0.1</td>
        <td>20</td>
        <td>0.0005400180816650391</td>
        <td>0.0002541542053222656</td>
        <td>0.07307142628533342</td>
        <td>0.10476153557145784</td>
    </tr>
</table>
